File: cSharp/Interfaces/DQN/InputVector.cs
using static Tensorflow.Binding;
using Tensorflow;
using NumSharp;
using System;
using System.Collections.Generic;
using lean_pactheman_client;
using PacTheMan.Models;
using System.Linq;

namespace lean_pactheman_client.DQN {
    public class InputVector {
        int selfPosition_X { get; set; }
        int selfPosition_Y { get; set; }

        int blinkyPosition_X { get; set; }
        int blinkyPosition_Y { get; set; }

        int inkyPosition_X { get; set; }
        int inkyPosition_Y { get; set; }

        int pinkyPosition_X { get; set; }
        int pinkyPosition_Y { get; set; }

        int clydePosition_X { get; set; }
        int clydePosition_Y { get; set; }

        public Tensor Tensor {
            get => new NDArray(
                    new float[,] {
                        {selfPosition_X, selfPosition_Y,
                        blinkyPosition_X, blinkyPosition_Y,
                        inkyPosition_X, inkyPosition_Y,
                        pinkyPosition_X, pinkyPosition_Y,
                        clydePosition_X, clydePosition_Y}
                    }
                );
        }

        public InputVector() {
            var selfDown = GameState.Instance.PlayerState.PlayerPositions[(Guid)GameState.Instance.Session.ClientId].Copy().Downscaled();
            selfPosition_X = (int)selfDown.X;
            selfPosition_Y = (int)selfDown.Y;

            var gPositions = GameState.Instance.GhostPositions.ToDictionary(kv => kv.Key, kv => kv.Value.Copy().Downscaled());
            blinkyPosition_X = (int)gPositions["blinky"].X;
            blinkyPosition_Y = (int)gPositions["blinky"].Y;

            inkyPosition_X = (int)gPositions["inky"].X;
            inkyPosition_Y = (int)gPositions["inky"].Y;

            pinkyPosition_X = (int)gPositions["pinky"].X;
            pinkyPosition_Y = (int)gPositions["pinky"].Y;

            clydePosition_X = (int)gPositions["clyde"].X;
            clydePosition_Y = (int)gPositions["clyde"].Y;
        }

        public InputVector(Position playerPosition, Dictionary<string, BasePosition> ghostPositions) {
            selfPosition_X = (int)playerPosition.Downscaled().X;
            selfPosition_Y = (int)playerPosition.Downscaled().Y;

            var gPositions = ghostPositions.ToDictionary(kv => kv.Key, kv => (kv.Value as Position).Copy().Downscaled());
            blinkyPosition_X = (int)gPositions["blinky"].X;
            blinkyPosition_Y = (int)gPositions["blinky"].Y;

            inkyPosition_X = (int)gPositions["inky"].X;
            inkyPosition_Y = (int)gPositions["inky"].Y;

            pinkyPosition_X = (int)gPositions["pinky"].X;
            pinkyPosition_Y = (int)gPositions["pinky"].Y;

            clydePosition_X = (int)gPositions["clyde"].X;
            clydePosition_Y = (int)gPositions["clyde"].Y;
        }
    }
}

File: typescript/src/interfaces/DQN/inputVector.ts
import GameState from "../../classes/gameState";
import { IPosition } from "../../models/pactheman.models";
import PositionExtended from "../../utils/extensions/positionExtensions";
import { tensor } from "@tensorflow/tfjs-node";

export default class InputVector {
    selfPosition_x: number;
    selfPosition_y: number;

    blinkyPosition_x: number;
    blinkyPosition_y: number;

    inkyPosition_x: number;
    inkyPosition_y: number;

    pinkyPosition_x: number;
    pinkyPosition_y: number;

    clydePosition_x: number;
    clydePosition_y: number;

    get asTensor() {
        return tensor([this.selfPosition_x, this.selfPosition_y, this.blinkyPosition_x, this.blinkyPosition_y,
        this.inkyPosition_x, this.inkyPosition_y, this.pinkyPosition_x, this.pinkyPosition_y,
        this.clydePosition_x, this.clydePosition_y], [1, 10]);
    }

    get tensorlike() {
        return [this.selfPosition_x, this.selfPosition_y, this.blinkyPosition_x, this.blinkyPosition_y,
        this.inkyPosition_x, this.inkyPosition_y, this.pinkyPosition_x, this.pinkyPosition_y,
        this.clydePosition_x, this.clydePosition_y];
    }

    get shape() {
        return [10];
    }

    constructor() {
        let selfDownscaled = PositionExtended.fromPosition(
            GameState.Instance.playerState.playerPositions.get(GameState.Instance.session.clientId ?? "")
            ?? { x: 0, y: 0 }).downscaled();
        this.selfPosition_x = selfDownscaled.x;
        this.selfPosition_y = selfDownscaled.y;

        let blinky = PositionExtended.fromPosition(GameState.Instance.ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        this.blinkyPosition_x = blinky.x;
        this.blinkyPosition_y = blinky.y;

        let inky = PositionExtended.fromPosition(GameState.Instance.ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        this.inkyPosition_x = inky.x;
        this.inkyPosition_y = inky.y;

        let pinky = PositionExtended.fromPosition(GameState.Instance.ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        this.pinkyPosition_x = pinky.x;
        this.pinkyPosition_y = pinky.y;

        let clyde = PositionExtended.fromPosition(GameState.Instance.ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        this.clydePosition_x = clyde.x;
        this.clydePosition_y = clyde.y;
    }

    static fromStateUpdate(selfPos: IPosition, ghostPositions: Map<string, IPosition>) {
        let vector = new InputVector();

        let selfDownscaled = PositionExtended.fromPosition(selfPos).downscaled();
        vector.selfPosition_x = selfDownscaled.x;
        vector.selfPosition_y = selfDownscaled.y;

        let blinky = PositionExtended.fromPosition(ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        vector.blinkyPosition_x = blinky.x;
        vector.blinkyPosition_y = blinky.y;

        let inky = PositionExtended.fromPosition(ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        vector.inkyPosition_x = inky.x;
        vector.inkyPosition_y = inky.y;

        let pinky = PositionExtended.fromPosition(ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        vector.pinkyPosition_x = pinky.x;
        vector.pinkyPosition_y = pinky.y;

        let clyde = PositionExtended.fromPosition(ghostPositions.get("blinky") ?? { x: 0, y: 0 }).downscaled();
        vector.clydePosition_x = clyde.x;
        vector.clydePosition_y = clyde.y;

        return vector;
    }
}